website/src/documentation/transforms/python/element-wise/pardo.md

```diff
@@ -19,26 +19,180 @@ limitations under the License.
+        width="20px" height="20px" alt="View on GitHub" />
+      View on GitHub
+    </a>
+  </td>
+</table>
+<br>
+
+### Example 3: ParDo with DoFn methods
+
+A [`DoFn`](https://beam.apache.org/releases/pydoc/current/apache_beam.transforms.core.html#apache_beam.transforms.core.DoFn)
+can be customized with a number of methods that can help create more complex behaviors.
+You can customize what a worker does when it starts and shuts down with `setup` and `teardown`.
+You can also customize what to do when a
+[*bundle of elements*](https://beam.apache.org/documentation/execution-model/#bundling-and-persistence)
+starts and finishes with `start_bundle` and `finish_bundle`.
+
+* [`DoFn.setup()`](https://beam.apache.org/releases/pydoc/current/apache_beam.transforms.core.html#apache_beam.transforms.core.DoFn.setup):
+  Called *once per worker* when the worker is starting to run.
+  This is a good place to connect to database instances, open network connections or other resources.
+
+* [`DoFn.start_bundle()`](https://beam.apache.org/releases/pydoc/current/apache_beam.transforms.core.html#apache_beam.transforms.core.DoFn.start_bundle):
+  Called *once per bundle of elements* before calling `process` on the first element of the bundle.
+  This is a good place to start keeping track of the bundle elements.
+
+* [**`DoFn.process(element, *args, **kwargs)`**](https://beam.apache.org/releases/pydoc/current/apache_beam.transforms.core.html#apache_beam.transforms.core.DoFn.process):
+  Called *once per element*, can *yield zero or more elements*.
+  Additional `*args` or `**kwargs` can be passed through
+  [`beam.ParDo()`](https://beam.apache.org/releases/pydoc/current/apache_beam.transforms.core.html#apache_beam.transforms.core.ParDo).
+  **[required]**
+
+* [`DoFn.finish_bundle()`](https://beam.apache.org/releases/pydoc/current/apache_beam.transforms.core.html#apache_beam.transforms.core.DoFn.finish_bundle):
+  Called *once per bundle of elements* after calling `process` after the last element of the bundle,
+  can *yield zero or more elements*. This is a good place to do batch calls on a bundle of elements,
+  such as running a database query.
+
+  For example, you can initialize a batch in `start_bundle`,
+  add elements to the batch in `process` instead of yielding them,
+  then running a batch query on those elements on `finish_bundle`, and yielding all the results.
+
+  Note that yielded elements from `finish_bundle` must be of the type
+  [`apache_beam.utils.windowed_value.WindowedValue`](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/utils/windowed_value.py).
+  You need to provide a timestamp as a unix timestamp, which you can get from the last processed element.
+  You also need to provide a window, which you can get from the last processed element like in the example below.
+
+* [`DoFn.teardown()`](https://beam.apache.org/releases/pydoc/current/apache_beam.transforms.core.html#apache_beam.transforms.core.DoFn.teardown):
```

Comment:
Teardown is called best effort and not guaranteed. For example, if worker crashes it may not be called.

Reply:
Thanks, clarified as well